Implementation of TaskChain

Redrield opened the pull request:
First, an overview of what TaskChain is
TaskChain is a library built by Aikar to make switches between sync and async less painful. The one notable point to this is that it allows data transiency without the nested executors that Sponge provides

In the commands where there are executors that are nested, these calls have been replaced by a TaskChain that works in the same way.

Do note that because of these changes the jar has to be created with `gradle build shadowJar` in order for dependencies to be shaded

Commit:
Optimized various parts of plugin

Commit:
Replace nested executor declarations with TaskChain (Added dependency and gradle shadow plugin for this to work. In order to build a working jarfile, files should be created with `gradle build shadowJar`

Commit:
Added name to authors list

parlough commented on src/main/java/com/meronat/iplog/commands/HistoryCommand.java:
```diff
@@ -57,7 +57,7 @@ public CommandResult execute(CommandSource src, CommandContext args) throws Comm
         Optional<InetAddress> optionalAddress = args.getOne("ip");
 
-        if (optionalUser.isPresent() && optionalAddress.isPresent()) {
+        if (!optionalUser.isPresent() && !optionalAddress.isPresent()) {
```
This exception is if they specify both so I believe there should be not `!` for either of these, as this case is handled in the `else` on line 127 below.

Redrield replied:
Alright, wasn't aware. I'll remove that

parlough commented on src/main/java/com/meronat/iplog/storage/Storage.java:
```diff
@@ -81,8 +81,8 @@ private void createTables() throws SQLException {
             conn.prepareStatement("CREATE TABLE IF NOT EXISTS REGISTRY("
-                + " IP VARCHAR(45),"
-                + " ID VARCHAR(36),"
+                + " IP CHAR(45),"
```
It's smart moving ID to `CHAR` but my understanding is that IP should be a `VARCHAR` still(if I am representing it as a string which I am) as IPv4 addresses are less than 45 characters and variable while I believe IPv6 addresses can be anywhere from 39 to 45 characters in length. I don't know SQL that well though so I may be wrong.

Redrield replied:
Hm, wasn't aware of that, I'll change IP back to VARCHAR

parlough commented on build.gradle:
```diff
@@ -14,8 +15,25 @@ compileJava.options.encoding = 'UTF-8'
+
+    shade "co.aikar:taskchain-sponge:3.4.3"
 }
 
```
Instead of changing the build statement required, couldn't we add something like the following to the build.gradle? I am not too experienced with Gradle so I'm not sure.

```
shadowJar {
    dependencies {
        include dependency('co.aikar:taskchain-sponge')
    }
}
```

parlough commented on build.gradle:
```diff
@@ -4,6 +4,7 @@ plugins {
     id 'signing'
     id 'org.spongepowered.plugin' version '0.8.1'
     id 'net.minecrell.licenser' version '0.3'
+    id "com.github.johnrengelman.shadow" version "1.2.3"
```
Change this to shadow version 1.2.4 as I believe that is out, unless there is a reason you used 1.2.3.

Redrield replied:
Wasn't aware that 1.2.4 was out, will do

parlough commented on build.gradle:
```diff
@@ -14,8 +15,25 @@ compileJava.options.encoding = 'UTF-8'
+repositories {
+    maven {
+        name = "aikar"
+        url = "http://ci.emc.gs/nexus/content/groups/aikar"
```
Does he have an https version of his maven repo?

Redrield replied:
No

Redrield replied:
Huh, apparently he does

parlough commented on src/main/java/com/meronat/iplog/IPLog.java:
```diff
@@ -185,10 +192,21 @@ public Logger getLogger() {
 
     }
 
+    public PluginContainer getPluginContainer() {
```
Do you plan to use this anywhere or is it just for convenience if we need it later?

Redrield replied:
I find it in good practice to have the PluginContainer available, because someone once told me that it was better practice to pass a PluginContainer to all the sponge methods with a parameter of Object plugin. As far as I remember he said that passing something else would just lead to a map lookup for the PluginContainer. Because of this I like to have it available for things like passing Cause etc. Not necessary, you can remove if you want.

parlough commented on src/main/java/com/meronat/iplog/IPLog.java:
```diff
@@ -185,10 +192,21 @@ public Logger getLogger() {
+    }
+
+    public static <T> TaskChain<T> newSharedChain(String name) {
+        return factory.newSharedChain(name);
```
I'm assuming you will use this somewhere?

Redrield replied:
Just read up on the topic, Aikar describes shared chains as something that you use when you want an operation to be nonblocking of the main thread, but it is blocking the object it's operating on. If at any point you feel like you would need this, then replace newChain calls in that scenario with newSharedChain(String) calls. If not, it can be removed

parlough reviewed (changes requested):
I really appreciate this pull request, I was not sure if Task Chain would be right for this plugin when I first looked, but with your pull request I see it will be fairly useful. There are a few suggestions/questions I have though. 

Thanks Redrield :D

parlough commented on src/main/java/com/meronat/iplog/commands/AddCommand.java:
```diff
@@ -72,11 +72,7 @@ public CommandResult execute(CommandSource src, CommandContext args) throws Comm
-            IPLog.getPlugin().getStorage().addConnection(ip, user.getUniqueId(), LocalDateTime.now());
-
-        });
+        Sponge.getScheduler().createAsyncExecutor(IPLog.getPlugin()).execute(() -> IPLog.getPlugin().getStorage().addConnection(ip, user.getUniqueId(), LocalDateTime.now()));
```
Could this be done to the purge command as well?

Redrield replied:
Yep

Commit:
Done modifications

parlough commented:
I will test it soon, then bar any complications I will merge it in. Thanks.

parlough committed:
Fixes a few things

Reverts build.gradle to more Red's original style. Fixes alias command to show name and not toString() representation of User, and changes lookup for address to use task chain as well. Also changes aikar's repository to https.

parlough commented:
I changed a few things like going back to how you setup the build.gradle as I had problems with how I suggested, added use of TaskChain to one of the lookup functions, and fixed a problem with the alias command.

Thank you for the pull request.
